app/controllers/products_controller.rb

```diff
@@ -0,0 +1,156 @@
+class ProductsController < ApplicationController
+  after_action :verify_authorized
+
+  # GET /products
+  def index
+    authorize Product
+    all_products = Product.all
+
+    # create two arrays with all the unique brand and material names that are used by the template products/index.html.erb to create html checkboxes
+    # the lists are also used by this function to filter products based on query params in the URL
+    @all_brands = Product.distinct.pluck(:brand)
+    @all_materials = Product.distinct.pluck(:material)
+
+    # create a list of all the products to be displayed by the in html file, after filtering it based on the query params in the URL
+    # params are filtered after "(brandX OR brandY) AND (materialA OR materialB) AND lprice AND hprice" logic
+    @products = Array.new
+    all_products.each do |product|
+      add = true
+      if params[:brand] != nil
+        add = false
+        brand_filter = paramsToList(params[:brand])
+        brand_filter.each do |brandNum|
+          if product.brand == @all_brands[brandNum]
+            add = true
+            break
+          end
+        end
+      end
+      if params[:material] != nil && add
+        add = false
+        material_filter = paramsToList(params[:material])
+        material_filter.each do |matNum|
+          if product.material == @all_materials[matNum]
+            add = true
+            break
+          end
+        end
+      end
+      if params[:lprice] != nil && priceStringToInt(params[:lprice]) != nil && priceStringToInt(params[:lprice]) > product.price
+        add = false
+      end 
+      if params[:hprice] != nil && priceStringToInt(params[:lprice]) != nil && priceStringToInt(params[:hprice]) < product.price
+        add = false
+      end
+      if params[:search] != nil && notMatchingSearchTerm(params[:search], product)
+        add = false
+      end
+      if add
+        @products.insert(-1, product)
```

Comment:
Q23

Are you sure you want to do the filtering in Ruby, and not in the database? This might not scale well as the number of products grows. See http://guides.rubyonrails.org/active_record_querying.html#conditions